Make ETW rundown operate more effectively with JIT

AaronRobinsonMSFT opened the pull request:
Fixes #102858

Make the `EECodeGenManager` API surface narrower.
Rename some locals removing critical section notion and replace with lock.

Rework the `CodeHeapIterator`.
Rename members to more accurately reflect their purpose. Generally move to more canonical C++.

`EECodeGenManager` now has an iterator counter. This is used to track the number of outstanding iterators and can be used to defer deletes. Adds are still valid since the iterator will only operate on the state at the time the iterator was created.

Always emit UnwindInfo for methods. This was previously only enabled when the Runtime provider was enabled and a rundown triggered.

AaronRobinsonMSFT committed:
Make the EECodeGenManager API surface narrower.
Rename some locals removing critical section notion
and replace with lock.

Rework the CodeHeapIterator.
Rename members to more accurately reflect their purpose.
Generally move to more canonical C++.

EECodeGenManager now has a iterator counter. This is
used to track the number of outstanding iterators and
can be used to defer deletes. Adds are still valid since
the iterator will only operate on the state at the time the
iterator was created.

AaronRobinsonMSFT committed:
Fix GCC build.

AaronRobinsonMSFT commented on src/coreclr/vm/class.cpp:
```diff
@@ -2892,56 +2892,38 @@ WORD SparseVTableMap::GetNumVTableSlots()
     if (pNewChunk->GetNextChunk() != NULL)
         return;
 
-    // even if pNewChunk->GetNextChunk() is NULL, this may still be the first chunk we added
-    // (last in the list) so find the end of the list and verify that
-    MethodDescChunk *chunk = GetChunks();
-    if (chunk != NULL)
-    {
-        while (chunk->GetNextChunk() != NULL)
-            chunk = chunk->GetNextChunk();
-
-        if (chunk == pNewChunk)
-            return;
-    }
-
-    pNewChunk->SetNextChunk(GetChunks());
-    SetChunks(pNewChunk);
```
This particular implementation is an inversion of what is in `AddChunk`, but I see no reason to retain that inversion. The inversion is the most recent `MethodDescChunk` is always first but for `AddChunk`, based on the comment there, the oldest is first.

AaronRobinsonMSFT replied:
This has been reverted.

AaronRobinsonMSFT committed:
Fix contracts

AaronRobinsonMSFT committed:
Fix contracts

AaronRobinsonMSFT committed:
Narrow contract

jkotas commented on src/coreclr/vm/profilingenumerators.cpp:
```diff
@@ -47,46 +47,51 @@ BOOL ProfilerFunctionEnum::Init(BOOL fWithReJITIDs)
 
-    EEJitManager::CodeHeapIterator heapIterator;
-    while(heapIterator.Next())
+    BOOL result = TRUE;
```
What changed that we need try/catch here now?

jkotas commented on src/coreclr/vm/crst.h:
```diff
@@ -352,17 +352,40 @@ class CrstBase
+        }
+
+        CrstHolder(CrstHolder const&) = delete;
+        CrstHolder &operator=(CrstHolder const&) = delete;
```
This looks like a re-implementation of `CrstHolderWithState` below. I would leave plain `CrstHolder` alone (see command "CrstHolder is optimized for the common use that takes the lock in constructor..." above). Change CrstHolderWithState to do what you need if it does not do it already.

AaronRobinsonMSFT replied:
I don't need this anymore actually. I'm going to make an explicit delete of the copy operators in question though. The rules for when they are implicitly marked delete is confusing.

jkotas reviewed (commented):
Test failures look related:
```
20:03:09.305 Running test: tracing/runtimeeventsource/nativeruntimeeventsource/nativeruntimeeventsource.cmd
ASSERT FAILED
	Expression: codeLength > 0
	Location:   line 1866 in /Users/runner/work/1/s/src/coreclr/vm/eetwain.cpp
```

AaronRobinsonMSFT committed:
Fix thread mode on DynamicMethod clean-up
Narrow contract

AaronRobinsonMSFT commented:
> Test failures look related:
> 
> ```
> 20:03:09.305 Running test: tracing/runtimeeventsource/nativeruntimeeventsource/nativeruntimeeventsource.cmd
> ASSERT FAILED
> 	Expression: codeLength > 0
> 	Location:   line 1866 in /Users/runner/work/1/s/src/coreclr/vm/eetwain.cpp
> ```

Yep. I've trying to root it out on macOS. I tried on Windows and did find a contract violation that I needed to fix. I pushed up that fix and will see where we are.

jkotas commented on src/coreclr/vm/runtimehandles.cpp:
```diff
@@ -1766,22 +1766,24 @@ extern "C" void QCALLTYPE RuntimeMethodHandle_Destroy(MethodDesc * pMethod)
+        END_PROFILER_CALLBACK();
+    }
 
     pDynamicMethodDesc->Destroy();
```
Is `DynamicMethodDesc::Destroy()` always called in pre-emptive mode now? If it is the case, it would be nice to make its contract more strict (`MODE_ANY` -> `MODE_PREEMPTIVE`) and delete unnecessary mode switch in the implementation.

AaronRobinsonMSFT committed:
Feedback

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -3713,16 +3822,44 @@ void EECodeGenManager::FreeCodeMemory(HostCodeHeap *pCodeHeap, void * codeStart)
+        // If we are in the middle of an enumeration, we cannot unload any code heap code.
+        EX_TRY
+        {
+            FreeHostCodeHeapStartCode* delayed = m_delayFreeHostCodeHeapMemory.AppendThrowing();
```
Is the code header going to point to DynamicMethodDesc that may get reused for other methods? Can it cause problems?

We may want to delay freeing the whole DynamicMethodDescs.

AaronRobinsonMSFT replied:
This is done.

AaronRobinsonMSFT committed:
Fix clean-up of delay structures.

AaronRobinsonMSFT committed:
Move DynamicMethodDesc destruction to finalizer queue if not
possible due to MethodDesc iteration.
Rework how DynamicMethodDesc and LCGMethodResolver
destruction works. This was needed due to Crst lock inversion.
Remove CrstIbcProfile

AaronRobinsonMSFT committed:
Fix contract

jkotas commented on src/coreclr/vm/class.h:
```diff
@@ -1694,7 +1698,12 @@ class EEClass // DO NOT CREATE A NEW EEClass USING NEW!
+    // after being set. The "current" pointer will be updated as new MethodDescChunks
+    // are added.
+    PTR_MethodDescChunk m_pChunksBegin;
+    PTR_MethodDescChunk m_pChunksCurrent;
```
Can we add new chunks at the beginning instead of adding this pointer here?

The order of these chunks should not matter.

jkotas replied:
Ah ok, I forgot that we introduced a dependency from reflection in https://github.com/dotnet/runtime/pull/69506/files#diff-4cae1ac340986fa73c94616dbe7cdb205ea0c9bea8d2adc5671a2ca3f00d9ffaR2714-R2717

jkotas commented on src/coreclr/vm/class.h:
```diff
@@ -1387,13 +1387,17 @@ class EEClass // DO NOT CREATE A NEW EEClass USING NEW!
+        _ASSERTE(pChunks != NULL);
+        _ASSERTE(m_pChunksBegin == NULL);
+        _ASSERTE(m_pChunksCurrent == NULL);
+        m_pChunksBegin = pChunks;
```
I do not think that this is correct. `MethodDescChunk* pChunks` can be a linked list of chunks. so this would have to walk through the linked list to find the end to compute the current chunk.

Is this refactoring related the rest of the changes? I think it would be best to do it separately. There are a lot of subtle details on how it works.

I assume that this change was motivated by `AddChunk` call from dynamicmethod.cpp since it does more than necesary. I would be ok with introducing `AddChunkUnordered` that just adds the new chunk to the start of the linked list and call it from dynamicmethod.cpp. The order for dynamic methods does not matter, so it is easy to reason about. But I would drop the larger refactoring that introduces the second pointer.

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -438,34 +443,26 @@ void UnwindInfoTable::AddToUnwindInfoTable(UnwindInfoTable** unwindInfoPtr, PT_R
+                CodeHeader * pHeader = pJitMgr->GetCodeHeaderFromStartAddress(methodEntry);
+                int unwindInfoCount = pHeader->GetNumberOfUnwindInfos();
+                for(int i = 0; i < unwindInfoCount; i++)
+                    AddToUnwindInfoTable(&pRS->_pUnwindInfoTable, pHeader->GetUnwindInfo(i), pRS->_range.RangeStart(), pRS->_range.RangeEndOpen());
```
Can we end up seeing a method that was created after the iteration started and end up publishing the unwind info again? Is the system going to handle the republishing gracefully?

One possible way to handle this is to publish these unwind infos unconditionally all the time. They are likely published most of the time anyway since there is typically some tracing session enable, so we are not saving much by publishing them on demand.

AaronRobinsonMSFT replied:
> Can we end up seeing a method that was created after the iteration started and end up publishing the unwind info again? Is the system going to handle the republishing gracefully?

I don't think so. The iterator here captures a snapshot of the current range and only examines that range of data. See `CodeHeapIterator` constructor in `codeman.cpp` for where I'm capturing the current state. Perhaps I'm missing something here.

jkotas replied:
I am worried about the following sequence:

- Iteration starts
- Dynamic methods get created and reuses space in one of the existing heaps
- Unwind info for this dynamic method gets published here
- Iteration reaches the heap where dynamic method reused space in step 2
- The unwind gets published again

Is there something that prevents this sequence from happening?

jkotas replied:
As far as I can tell, https://github.com/dotnet/runtime/pull/116354/files#diff-eba17a72b6caa8b52b196f3b84bfd20dfeaf39a89f410943a692d1f33f569f3bR536 does not prevent reuse of the free space that's within the range for dynamic methods

AaronRobinsonMSFT replied:
Nevermind. I see what you are suggesting here. I assumed this function was doing something else.

AaronRobinsonMSFT replied:
Done.

AaronRobinsonMSFT committed:
Remove GetChunk optimization.

AaronRobinsonMSFT committed:
Move finalizer related work functions to finalizerthread.cpp.

AaronRobinsonMSFT committed:
Alter HostCodeHeap memory clean up.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into runtime_102858

AaronRobinsonMSFT committed:
Nits

AaronRobinsonMSFT committed:
Unconditionally publish UnwindInfo.

AaronRobinsonMSFT committed:
Missed function rename.

AaronRobinsonMSFT committed:
Contract violation fix
Fixup Crst relationship of WrapperTemplate and ExecutableAllocatorLock.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into runtime_102858

AaronRobinsonMSFT commented:
@jkotas I think this should be ready to review now.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into runtime_102858

jkotas commented on src/coreclr/vm/ceemain.cpp:
```diff
@@ -755,6 +755,13 @@ void EEStartupHelper()
         InitEventStore();
 
+#ifdef TARGET_AMD64
+        // We only do this on amd64 (NOT ARM, because ARM uses frame based stack crawling),
```
We don't do this on non-Windows either, but the ifdefs are in different spot. I think it would look best to have a no-op implementation of `UnwindInfoTable` for non-Amd64 non-Windows so that these ifdefs are not spread around.

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -148,10 +149,9 @@ bool InitUnwindFtns()
         }
         RtlUnwindFtnsInited = true;
```
`RtlUnwindFtnsInited` can be deleted now that we initialize this during startup

AaronRobinsonMSFT committed:
Feedback

jkotas committed:
Merge branch 'main' into runtime_102858

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -3280,74 +3241,180 @@ BYTE* EECodeGenManager::allocFromJitMetaHeap(MethodDesc* pMD, DWORD blockSize, s
+        EX_CATCH
+        {
+            MethodDesc* pMD = arg.GetMethodDesc();
+            LOG((LF_JIT, LL_ERROR, "Failed to delay RemoveJitData for method, %s::%s\n", pMD->m_pszDebugClassName, pMD->m_pszDebugMethodName));
```
I am not sure whether dealing with RemoveJitData is worth the trouble:

For !IsLCGMethod case, the backout for `GetJitMetaHeap` looks like overengineering given that we are not backing out the code itself that is by far the largest chunk of memory. I think it would be fine to delete `GetJitMetaHeap` backout logic to keep this simpler.

It leaves is with the IsLCGMethod case. If we hit this OOM path, we will end up with code allocated for DynamicMethod method, but that DynamicMethod will be eventually destroyed. I am not sure whether all parts of the system will be able to handle this stray piece of code heap memory attached to destroyed DynamicMethod gracefully. It may be better to turn `LCGMethodResolver::m_recordCodePointer` into a linked list, same way as e.g. `LCGMethodResolver::m_DynamicStringLiterals` are done, and do cleanup of all code blocks as part of the regular DynamicMethod cleanup. As side effect, it would prepare us for enabling tiered compilation for DynamicMethods. DynamicMethods do not support tiered compilation because of it is not possible to attach multiple pieces of code to a single DynamicMethod.  

If we do both of these, the entire `RemoveJitData` can be deleted.

AaronRobinsonMSFT replied:
Let me see what I can do here.

AaronRobinsonMSFT committed:
Feedback

AaronRobinsonMSFT commented on src/coreclr/vm/finalizerthread.cpp:
```diff
@@ -46,11 +46,114 @@ void FinalizerThread::EnableFinalization()
+        STANDARD_VM_CONTRACT;
+        _ASSERTE(pDMD != NULL);
+
+        s_delayDestroyDynamicMethodList.LinkHead(new SimpleList<DynamicMethodDesc*>::Node(pDMD));
```
I disagree in this case. This allocation is on the native heap and the games we tend to play with the `m_next` pattern are often far more complex than needed. I will give this a shot, but all it is going to do is create an even more complex clean-up scenario for an already niche case.

AaronRobinsonMSFT replied:
> There is convenient LCGMethodResolver::m_next that we can reuse for this link list.

This seems to be already used for some clean-up scenario. Sigh.... detangling this isn't really something I'm interested in doing, but I'll give it a shot.

AaronRobinsonMSFT commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -3518,14 +3564,48 @@ void EEJitManager::DeleteFunctionTable(PVOID pvTableID)
+    if (m_iteratorCount != 0)
+    {
+        // Record the LoaderAllocator and return.
+        LoaderAllocator** toUnload = m_delayUnload.AppendThrowing();
```
I will try this too, but I tend to consider the `m_next` pattern of lists on various types a code smell. We don't use RAII correctly, we don't clean-up in a predictable way and finally we don't follow consistent patterns for this. Which means by using this approach we create a system that is very fragile to work with for the benefit of a niche scenarios. I would prefer to keep clean-up clear and reliable so we can improve slowly rather than use patterns that were deemed unsafe decades ago after C++ was introduced.

AaronRobinsonMSFT replied:
> as long as it is non-throwing
> Is there the right way (tm) to do this in modern C++?

Not sure non-throwing and modern C++ align. The non-throwing part is the sticking point for me. If there is no more native memory there isn't much we can do and trying to address that rarely works. For example, Swift doesn't even try, it just fails if I recall. I get this isn't true from the managed memory perspective, but if we are using native memory and there isn't any to provide, there isn't much we can do and any fallback is likely to be a temporary patch for a moment. This is particularly true in a clean-up path.

As I said, I'm going to attempt to remove the allocation, but I'm not convinced the benefit is worth the complexity.

jkotas replied:
> For example, Swift doesn't even try,

I would be ok with fail fasting on allocation failure here. We do that in a few places.

AaronRobinsonMSFT committed:
Nits

AaronRobinsonMSFT committed:
Handle the OOM case for clean-up

AaronRobinsonMSFT committed:
Remove RemoveJitData() logic.

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -5422,7 +5340,7 @@ void ExecutionManager::EnumMemoryRegions(CLRDataEnumMemoryFlags flags)
 {
     CONTRACTL {
-        NOTHROW;
+        THROWS;
```
Can this be NOTHROW now?

AaronRobinsonMSFT replied:
Yes.

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -2898,15 +2844,15 @@ void EECodeGenManager::allocCode(MethodDesc* pMD, size_t blockSize, size_t reser
         if (pMD->IsLCGMethod())
         {
-            pMD->AsDynamicMethodDesc()->GetLCGMethodResolver()->m_recordCodePointer = (void*) pCode;
+            pMD->AsDynamicMethodDesc()->GetLCGMethodResolver()->RecordCodePointer((void*)pCode);
```
Allocate the linked list note for the recording before the call to `AllocCodeWorker` so that `RecordCodePointer` cannot fail and leave a piece of memory that's not attached to anything behind?

AaronRobinsonMSFT replied:
Sure.

jkotas commented on src/coreclr/vm/dynamicmethod.h:
```diff
@@ -179,26 +184,27 @@ class LCGMethodResolver : public DynamicResolver
-    void *m_recordCodePointer;
+    DynamicMethodTable* m_DynamicMethodTable;
+    DynamicMethodDesc* m_next;
+    DynamicMethodDesc* m_delayDestroyNext;
```
Was there an actual problem with reusing `m_next` for the delay destroy list? I do not see by just looking at the code.

AaronRobinsonMSFT replied:
Yes. The issue is the `m_next` is still valid while it is delayed. It is part of the iteration chain hanging off of the `DynamicMethodTable`. In order to fix that up I would need to remove if from that chain. This is an example of how bad this pattern becomes as it confuses the element itself with how it is apart of a collection. Those are orthogonal concepts and shouldn't be mixed.

jkotas replied:
> It is part of the iteration chain hanging off of the DynamicMethodTable

What is this iteration chain that's hanging off of the DynamicMethodTable? I do not see it. As far as I can tell, `m_next` is only used for free list hanging off of `DynamicMethodTable::m_DynamicMethodList`. When the method is on the free list, it should not be in the delayed list. If `m_next` participates in a linked list that is not this free list, could you please add comment about it next to `m_next`?

> This is an example of how bad this pattern becomes as it confuses the element itself with how it is apart of a collection. 

I understand that you do not like this. On the other hand, linked lists like this - where `_next is` used to form different linked list depending on the state that the item is in - are very common in low-level programming (e.g. OSes).

jkotas replied:
> I don't believe we can nor should we try to be resilient on out of native memory scenarios

I do not think there is a difference between native memory OOM and managed memory OOM. The two types of memory are used interchangeably. I think having each type of memory on different plan would be hard to reason about.

I see our OOM robustness as good-enough. I agree that it would be nice to have explicit testing for it, but that's quite expensive project (we used to have this type of testing in .NET Framework). We regularly fix bugs in OOM handling based on code inspection, static analysis or customer feedback. For example, #116550 that's included in the current servicing payload (and that was hit by a customer) is fixing OOM robustness bug.

Our users are used to OOM turning into catchable OutOfMemory exceptions. In very rare situation where the cleanup path is too complicated, it is ok to turn OOM into failfast. I think it would be a tough breaking change to push through to do failfast for all OOMs.

AaronRobinsonMSFT commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10998,13 +10998,7 @@ void CEEJitInfo::BackoutJitData(EECodeGenManager * jitMgr)
 
-    // The RemoveJitData call below requires the m_CodeHeader to be valid, so we need to write
-    // the code bytes to the target memory location.
     WriteCodeBytes();
```
I can't justify this function based on current usage. I'll remove it.

AaronRobinsonMSFT committed:
Remove CEECodeGenInfo::BackoutJitData()
Rework allocation order for recording of code pointer.

jkotas commented on src/coreclr/vm/dynamicmethod.cpp:
```diff
@@ -997,15 +992,71 @@ void LCGMethodResolver::RecycleIndCells()
+    while (m_DynamicCodePointers != NULL)
+    {
+        // Remove the unwind information (if applicable)
+        void* recordCodePointer = m_DynamicCodePointers->m_pEntry;
```
This can be null if the actual memory allocation failed. Do we need to skip null entries here?

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into runtime_102858

AaronRobinsonMSFT committed:
Feedback.

AaronRobinsonMSFT commented on src/coreclr/vm/eventtrace.cpp:
```diff
@@ -3620,7 +3611,7 @@ VOID ETW::MethodLog::DynamicMethodDestroyed(MethodDesc *pMethodDesc)
     CONTRACTL {
         NOTHROW;
-        GC_TRIGGERS;
+        GC_NOTRIGGER;
```
Yep, this has been validated.

jkotas reviewed (approved):
LGTM. Thank you!
